File: spikes/platform/runtime/go/lease.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// leaseRecord is the JSON structure stored in the lease file.
type leaseRecord struct {
	Owner      string `json:"owner"`
	ExpiresUTC string `json:"expires_utc"`
}

// Lease is an exclusive advisory lease backed by a JSON file.
// The file stores {"owner": "<hex>", "expires_utc": "<RFC3339Nano>"}.
type Lease struct {
	path   string
	ttlMs  int
	owner  string // empty when not held
}

// newLease creates a Lease for the given file path and TTL.
func newLease(path string, ttlMs int) *Lease {
	return &Lease{path: path, ttlMs: ttlMs}
}

// Owner returns the owner hex string for this lease instance, or "".
func (l *Lease) Owner() string { return l.owner }

// generateOwner returns 16 random bytes encoded as a 32-char hex string.
func generateOwner() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", fmt.Errorf("generate owner: %w", err)
	}
	return hex.EncodeToString(b), nil
}

// expiry returns the expiry time for a new lease record.
func (l *Lease) expiry() time.Time {
	return time.Now().UTC().Add(time.Duration(l.ttlMs) * time.Millisecond)
}

// encode serialises the lease payload.
func (l *Lease) encode() ([]byte, error) {
	rec := leaseRecord{
		Owner:      l.owner,
		ExpiresUTC: l.expiry().Format(time.RFC3339Nano),
	}
	return json.Marshal(rec)
}

// readRecord reads the lease file and returns the parsed record.
// Returns (nil, nil) when the file does not exist.
// Returns a zero-value record with an error on parse failures.
func (l *Lease) readRecord() (*leaseRecord, error) {
	data, err := os.ReadFile(l.path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	var rec leaseRecord
	if err := json.Unmarshal(data, &rec); err != nil {
		return nil, fmt.Errorf("parse lease: %w", err)
	}
	if rec.Owner == "" || rec.ExpiresUTC == "" {
		return nil, fmt.Errorf("malformed lease: missing fields")
	}
	return &rec, nil
}

// isExpired reports whether the lease record has passed its expiry time.
func isExpired(rec *leaseRecord) bool {
	exp, err := time.Parse(time.RFC3339Nano, rec.ExpiresUTC)
	if err != nil {
		// Malformed expiry → treat as expired so we don't block forever.
		return true
	}
	return time.Now().UTC().After(exp)
}

// Acquire tries to acquire the lease. Returns true on success, false if busy.
func (l *Lease) Acquire() (bool, error) {
	if err := os.MkdirAll(filepath.Dir(l.path), 0o700); err != nil {
		return false, fmt.Errorf("acquire: mkdir: %w", err)
	}

	owner, err := generateOwner()
	if err != nil {
		return false, err
	}

	rec := leaseRecord{
		Owner:      owner,
		ExpiresUTC: l.expiry().Format(time.RFC3339Nano),
	}
	payload, err := json.Marshal(rec)
	if err != nil {
		return false, fmt.Errorf("acquire: marshal: %w", err)
	}

	// O_CREATE|O_EXCL — atomic exclusive create.
	f, err := os.OpenFile(l.path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o600)
	if err == nil {
		// Success: write payload and close.
		if _, werr := f.Write(payload); werr != nil {
			f.Close()
			os.Remove(l.path)
			return false, fmt.Errorf("acquire: write: %w", werr)
		}
		if serr := f.Sync(); serr != nil {
			f.Close()
			os.Remove(l.path)
			return false, fmt.Errorf("acquire: sync: %w", serr)
		}
		f.Close()
		l.owner = owner
		return true, nil
	}

	if !errors.Is(err, os.ErrExist) {
		return false, fmt.Errorf("acquire: open: %w", err)
	}

	// File exists. Read and check expiry.
	existing, rerr := l.readRecord()
	if rerr != nil || existing == nil {
		// Unreadable or gone — busy (or retry on gone, but let's be safe).
		return false, nil
	}
	if !isExpired(existing) {
		return false, nil // busy
	}

	// Expired — steal it.
	if rerr := os.Remove(l.path); rerr != nil && !errors.Is(rerr, os.ErrNotExist) {
		return false, nil // couldn't remove; treat as busy
	}
	return l.Acquire() // retry after removal
}

// Renew refreshes the lease expiry. Only succeeds if this instance owns it.
func (l *Lease) Renew() (bool, error) {
	if l.owner == "" {
		return false, nil
	}
	rec, err := l.readRecord()
	if err != nil || rec == nil || rec.Owner != l.owner {
		return false, nil
	}
	payload, err := l.encode()
	if err != nil {
		return false, fmt.Errorf("renew: encode: %w", err)
	}
	if err := atomicReplace(l.path, payload); err != nil {
		return false, fmt.Errorf("renew: atomic replace: %w", err)
	}
	return true, nil
}

// Release removes the lease file if this instance owns it.
func (l *Lease) Release() (bool, error) {
	if l.owner == "" {
		return false, nil
	}
	rec, err := l.readRecord()
	if err == nil && rec != nil && rec.Owner == l.owner {
		if rerr := os.Remove(l.path); rerr != nil && !errors.Is(rerr, os.ErrNotExist) {
			l.owner = ""
			return false, fmt.Errorf("release: remove: %w", rerr)
		}
		l.owner = ""
		return true, nil
	}
	l.owner = ""
	return false, nil
}

// IsActive returns true when the lease file exists (regardless of owner).
func (l *Lease) IsActive() bool {
	_, err := os.Stat(l.path)
	return err == nil
}

File: spikes/platform/runtime/dotnet/Lease.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Kinglet.HostProbe;

/// <summary>
/// File-backed exclusive advisory lease.
/// The file stores {"owner":"<hex>","expires_utc":"<ISO8601>"}.
/// </summary>
internal sealed class Lease
{
    private readonly string _path;
    private readonly int _ttlMs;
    private string? _owner; // non-null when this instance holds the lease

    public Lease(string path, int ttlMs)
    {
        _path = path;
        _ttlMs = ttlMs;
    }

    public string? Owner => _owner;

    // ------------------------------------------------------------------
    // Public API
    // ------------------------------------------------------------------

    /// <summary>Tries to acquire the lease. Returns true on success, false if busy.</summary>
    public bool Acquire()
    {
        Directory.CreateDirectory(Path.GetDirectoryName(_path)!);

        string owner = GenerateOwner();
        var record = new LeaseRecord(owner, Expiry());
        byte[] payload = Serialize(record);

        // O_CREATE | O_EXCL — atomic exclusive create.
        FileStream? fs = TryCreateExclusive(_path);
        if (fs != null)
        {
            using (fs)
            {
                fs.Write(payload);
                fs.Flush(flushToDisk: true);
            }
            _owner = owner;
            return true;
        }

        // File exists — check expiry.
        LeaseRecord? existing = TryReadRecord(_path);
        if (existing == null)
        {
            // Unreadable or gone — treat as busy.
            return false;
        }
        if (!IsExpired(existing))
        {
            return false; // actively held
        }

        // Expired — steal it.
        try { File.Delete(_path); } catch (FileNotFoundException) { }
        catch { return false; }

        return Acquire(); // retry
    }

    /// <summary>Refreshes the lease expiry. Only succeeds if this instance owns it.</summary>
    public bool Renew()
    {
        if (_owner == null) return false;
        LeaseRecord? rec = TryReadRecord(_path);
        if (rec == null || rec.Owner != _owner) return false;

        var updated = new LeaseRecord(_owner, Expiry());
        AtomicReplace(_path, Serialize(updated));
        return true;
    }

    /// <summary>Removes the lease file if this instance owns it.</summary>
    public bool Release()
    {
        if (_owner == null) return false;
        LeaseRecord? rec = TryReadRecord(_path);
        if (rec != null && rec.Owner == _owner)
        {
            try
            {
                File.Delete(_path);
                _owner = null;
                return true;
            }
            catch (FileNotFoundException)
            {
                _owner = null;
                return true;
            }
            catch
            {
                _owner = null;
                return false;
            }
        }
        _owner = null;
        return false;
    }

    /// <summary>Returns true if the lease file exists (regardless of owner).</summary>
    public bool IsActive() => File.Exists(_path);

    // ------------------------------------------------------------------
    // Internal helpers
    // ------------------------------------------------------------------

    private DateTime Expiry() => DateTime.UtcNow.AddMilliseconds(_ttlMs);

    private static bool IsExpired(LeaseRecord rec)
    {
        return DateTime.UtcNow > rec.ExpiresUtc;
    }

    private static string GenerateOwner()
    {
        Span<byte> bytes = stackalloc byte[16];
        RandomNumberGenerator.Fill(bytes);
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    private static byte[] Serialize(LeaseRecord record)
    {
        return JsonSerializer.SerializeToUtf8Bytes(record, LeaseJsonContext.Default.LeaseRecord);
    }

    private static LeaseRecord? TryReadRecord(string path)
    {
        try
        {
            byte[] data = File.ReadAllBytes(path);
            var rec = JsonSerializer.Deserialize(data, LeaseJsonContext.Default.LeaseRecord);
            if (rec == null || rec.Owner.Length == 0) return null;
            return rec;
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Opens path exclusively with FileMode.CreateNew. Returns null if it already exists.
    /// </summary>
    private static FileStream? TryCreateExclusive(string path)
    {
        try
        {
            return new FileStream(path, FileMode.CreateNew, FileAccess.Write, FileShare.None);
        }
        catch (IOException)
        {
            return null;
        }
    }

    // ------------------------------------------------------------------
    // Atomic replace (used by Renew)
    // ------------------------------------------------------------------

    internal static void AtomicReplace(string target, byte[] data)
    {
        string dir = Path.GetDirectoryName(target)!;
        Directory.CreateDirectory(dir);
        string tmp = target + "." + Guid.NewGuid().ToString("N") + ".tmp";
        try
        {
            using (var fs = new FileStream(tmp, FileMode.CreateNew, FileAccess.Write, FileShare.None))
            {
                fs.Write(data);
                fs.Flush(flushToDisk: true);
            }
            File.Move(tmp, target, overwrite: true);
        }
        catch
        {
            try { File.Delete(tmp); } catch { }
            throw;
        }
    }
}

// ------------------------------------------------------------------
// JSON model for lease file
// ------------------------------------------------------------------

internal sealed class LeaseRecord
{
    [JsonPropertyName("owner")]
    public string Owner { get; set; } = "";

    [JsonPropertyName("expires_utc")]
    public DateTime ExpiresUtc { get; set; }

    public LeaseRecord() { }

    public LeaseRecord(string owner, DateTime expiresUtc)
    {
        Owner = owner;
        ExpiresUtc = expiresUtc;
    }
}

[JsonSerializable(typeof(LeaseRecord))]
internal sealed partial class LeaseJsonContext : JsonSerializerContext { }
